### PR TITLE
link: allow creating public link to files and folders - closes #1562

### LINKED ISSUE #1562: Add command to generate download link
The latest version with the Windows mount is awesome, thanks a lot !

Can you add a command to generate HTTP download link when it's possible (not available when files are encrypted)

`./rclone link 'gdrive:/dir/myfile.avi'`

And it's generate / get the download link from the backend ?

I think adding this feature would not be very complicated, each backend already gets the file link from its API.

### PR DESCRIPTION
**Pull request by breunigs**
For now only Google Drive and Dropbox are included.

The goal of this command is to quickly get a sharable link to hand to friends. Drive supports full ACLs, Dropbox offers password protection and automatic expiry (paid accounts). Not sure if it makes sense to implement these provider specific features. The most likeliest feature to add would probably to "unshare", which could be equally universal.

It's an optional interface for `Fs`, although it would probably fit nicely on `Object` as well. As it is now, it supports sharing directories. Getting dir sharing to work via `Object` would likely benefit from having a standardized `Directory` type first. I haven't surveyed the other storage providers yet, so maybe demanding directory sharing through the integration tests will have to change.

Note that the Google Drive intergration test fails for me at `TestObjectRemove`, even on master.

There is one edge case in the Drive implementation that I left unsolved: on a team drive, if I lack permissions (i.e. `canShare: false`) this implementation will bail. Instead, it could check what permissions the file/dir already has, and return the link if they are already sufficient. I'd revisit this if someone actually runs into this issue.

As always, I am happy to address any feedback.

### CODE REVIEW

File: cmd/link/link.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    rclone link remote:path/to/folder/
+
+If successful, the last line of the output will contain the link. Exact
+capabilities depend on the remote, but the link will always be created with
+the least constraints – e.g. no expiry, no password protection, accessible
+without account.
+`,
+	Run: func(command *cobra.Command, args []string) {
+		cmd.CheckArgs(1, 1, command, args)
+		fsrc, srcFileName := cmd.NewFsSrcWithFile(args)
```
Comment (ncw) [inline cmd/link/link.go]: I would have thought you'd be better calling [newFsFile](https://github.com/ncw/rclone/blob/master/cmd/cmd.go#L147) and then you'll know whether the argument points to a file or directory without the need for a trailing `/` check which I've tried to avoid in rclone.  (You'd have to make it public first though...)

Then the PublicLink is then either sharing the root directory (if passed "") or a path within that directory.

I don't know - what do you think?

File: cmd/link/link.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+		cmd.Run(false, false, command, func() error {
+			link, err := operations.PublicLink(fsrc, srcFileName)
+			if err != nil {
+				return err
+			}
+			fmt.Printf("public link:\n%s\n", link)
```
Comment (ncw) [inline cmd/link/link.go]: This would probably be most useful for scripting if it is just the link without the `public link` text.

File: fs/operations/operations.go [hunk 3]
```diff
@@ -1470,6 +1470,15 @@ func Rcat(fdst fs.Fs, dstFileName string, in io.ReadCloser, modTime time.Time) (
 	return dst, nil
 }
 
+// PublicLink adds a "readable by anyone with link" permission on the given file or folder.
+func PublicLink(f fs.Fs, fileName string) (string, error) {
```
Comment (ncw) [inline fs/operations/operations.go]: This `fileName` should probably be `remote` and be relative to the Fs root as per normal rclone usage.

File: fs/fs.go [hunk 4]
```diff
@@ -642,6 +648,12 @@ type PutStreamer interface {
 	PutStream(in io.Reader, src ObjectInfo, options ...OpenOption) (Object, error)
 }
 
+// PublicLinker is an optional interface for Fs
+type PublicLinker interface {
+	// PublicLink generates a public link to the remote path (usually readable by anyone)
+	PublicLink(fileName string) (string, error)
```
Comment (ncw) [inline fs/fs.go]: This `fileName` should probably be `remote` and be relative to the Fs root as per normal rclone usage.

**Comment by ncw** [comment]
> Note that the Google Drive intergration test fails for me at TestObjectRemove, even on master.

I had noticed that, its been showing up in the [integration tests](https://pub.rclone.org/integration-tests/)...

I'll work in that in #2177

File: cmd/link/link.go (hunk 2 above)
Comment (breunigs) [reply cmd/link/link.go]: Just putting the link out there seemed wrong, and there always is `tail -n1`. On second thought the user pretty much expects a link, so the explanation is superfluous. I've removed it.

File: cmd/link/link.go (hunk 1 above)
Comment (breunigs) [reply cmd/link/link.go]: You are most certainly right, but I ran into issues previously and it was not clear to me where they were coming from. With Dropbox, I can simply drop in `NewFsFile` and it continues to work.

However, with Drive I run into the peculiar issue that the `f.root` *changes*. I added some debug output in the form of `fmt.Printf("%s @ %s  0\n", f.Root(), remote)` to the code. It's still fine in `operations.PublicLink`. It's fine at the top of Dropbox's `PublicLink`. However, looking at the debug output for Drive (`xxlink` is a file at the very root, i.e. no subroot):

```
$ go build -i && ./rclone -vvv link Gsuite:xxlink
2018/03/24 19:36:01 DEBUG : rclone: Version "v1.40-DEV" starting with parameters ["./rclone" "-vvv" "link" "Gsuite:xxlink"]
2018/03/24 19:36:01 DEBUG : Using config file from "/home/stefan/.rclone.conf"
 @ xxlink
 @ xxlink  0
xxlink @ xxlink  1
2018/03/24 19:36:02 DEBUG : Google drive root 'xxlink': attempting to share single file 'xxlink'
xxlink @ xxlink  2
drive fs: Google drive root 'xxlink'
2018/03/24 19:36:02 Failed to link: object not found
```

For comparison, this is the output for Dropbox (again `rclone139` is a top level file):

```
$ go build -i && ./rclone -vvv link dropbox:rclone139
2018/03/24 19:35:01 DEBUG : rclone: Version "v1.40-DEV" starting with parameters ["./rclone" "-vvv" "link" "dropbox:rclone139"]
2018/03/24 19:35:01 DEBUG : Using config file from "/home/stefan/.rclone.conf"
 @ rclone139
 @ rclone139  0
 @ rclone139  1
2018/03/24 19:35:02 DEBUG : Dropbox root '': attempting to share 'rclone139' (absolute path: /rclone139)
2018/03/24 19:35:02 DEBUG : /rclone139: has a public link already, attempting to retrieve it
https://www.dropbox.com/s/otsnp322yzdetup/rclone139?dl=0
2018/03/24 19:35:03 DEBUG : 4 go routines active
2018/03/24 19:35:03 DEBUG : rclone: Version "v1.40-DEV" finishing with parameters ["./rclone" "-vvv" "link" "dropbox:rclone139"]
```

So I guess there must be a bug somewhere in the drive code? Maybe `rootFolderId` or some of the dirCache is not cleaned properly… if you have a hunch, please let me know.

Comment (breunigs) [reply cmd/link/link.go]: Hm, it seems that the old or incorrect FS resurfaces somehow… I print the memory addresses in drive's `NewFs` implementation here: https://github.com/ncw/rclone/blob/master/backend/drive/drive.go#L571

```
in NewFs:
old: 0xc4203b4000 (f)
new: 0xc42057a090 (newF)

in operations.PublicLink:
0xc42057a090 (i.e. the correct one)

in drive.PublicLink:
0xc4203b4000 (i.e. the incorrect one)
```

so I guess `Features()` hangs on to the old reference somehow?

Comment (breunigs) [reply cmd/link/link.go]: So, it's exactly like that: the `f` features are filled with functions whose receiver is `*f`. These are copied when `newF` is created and are never updated – they keep `*f` as their receiver. I guess calling `.Fill()` again might work, but so does simply overwriting `f`'s `dirCache` and `root`. 

A test that ensures that the address of `f` matches the addresses of all receivers of `f.Features()` functions sounds like a decent idea. Unfortunately I could not find a way to get the pointer of a func's *receiver* from the func's own pointer.

**Comment by breunigs** [comment]
Writing a test case that ensures that `f` and `f.Features().someFeature`'s receiver match seems rather tough. A friend more versed in Go came up with this https://play.golang.org/p/so-l-ZSPCSo, but this is clearly not nice, even if we'd just use it in tests.

For the current patch, I'd like to merge it as it is now. Well, if you think it's helpful I can move drive's `NewFs` patch into a separate commit, so that bisecting it later might be easier.

In the long run I wonder though if `Features()` and the optional interfaces can be replaced by making every interface required and backends return `fs.ErrorNotImplemented` if they don't support an optional feature. I lack a grand overview of rclone to be able to judge if this is feasible. Do you see any roadblocks with this approach that make even trying not worth it? I am not sure this approach would be better, or if it's just "bad in a different kind of way", though.

**Comment by breunigs** [comment]
Split off the receiver issues into #2182. I will put the Drive patch there. Let's see if I can make a durable enough test, though!

**Comment by ncw** [comment]
Thanks @breunigs :-)
